File: atcoder/dp/d-knapsack-1.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
int main() {
    int n, W;
    long long ans = 0;
    cin >> n >> W;

    vector<bool> p(W + 1, false);  // constructed weight
    vector<int> w(n + 1, 0);       // item weight
    vector<long long> v(n + 1, 0); // item value
    vector<long long> V(W + 1, 0); // maximum value for constructed weight
    for (int i = 1; i <= n; i++) {
        cin >> w[i] >> v[i];
    }

    p[0] = true;
    for (int k = 1; k <= n; k++) {
        for (int x = W; x >= 0; x--) {
            if (p[x] == true && x + w[k] <= W) {
                p[x + w[k]] = true;
                V[x + w[k]] = max(V[x + w[k]], V[x] + v[k]);
                ans = max(ans, V[x + w[k]]);
            }
        }
    }
    cout << ans << endl;

    return 0;
}

File: atcoder/dp/D-Knapsack-1.py
N,W = map(int, input().split())
w = []
v = []
for _ in range(N):
    weight,value = map(int, input().split())
    w.append(weight)
    v.append(value)

# rows = items, cols = weights
# dp[i][j] = maximum value for weight j using first i items 
dp = [[0 for col in range(W+1)] for row in range(N+1)]

for i in range(1, N+1):
    for j in range(0, W+1):
        w_i = w[i-1]
        v_i = v[i-1]
        if j - w_i >= 0:
            dp[i][j] = max(dp[i-1][j], dp[i-1][j-w_i] + v_i)
        else:
            dp[i][j] = dp[i-1][j]
print(dp[N][W])
